core/src/crds_trie.rs

```diff
@@ -0,0 +1,359 @@
+use crate::crds::VersionedCrdsValue;
+use crate::crds_gossip_pull::CrdsFilter;
+use crate::crds_value::CrdsValueLabel;
+
+/// Most significant bit in u64.
+const MSB_U64: u64 = 1u64 << 63;
+
+#[derive(Clone)]
+pub enum CrdsTrie {
+    // Left subtree, size of subtree (num of values at leaves), right subtree.
+    // Size is used in pruning to drop empty branches.
+    Node(Link, usize, Link),
+    // Since values are placed at hashes, collisions should be rare, and so a
+    // vector is more efficient than a (hash) set.
+    Leaf(Vec<CrdsValueLabel>),
+}
+
+type Link = Option<Box<CrdsTrie>>;
+
+fn new_link(leaf: bool) -> Link {
+    let node = if leaf {
+        CrdsTrie::Leaf(vec![])
+    } else {
+        CrdsTrie::Node(None, 0, None)
+    };
+    Some(Box::<CrdsTrie>::new(node))
+}
+
+/// Unboxes a &Option<Box<CrdsTrie>> to Option<&CrdsTrie>.
+#[inline]
+fn unbox_link(link: &Link) -> Option<&CrdsTrie> {
+    match link {
+        None => None,
+        Some(node) => Some(&*node),
+    }
+}
+
+impl Default for CrdsTrie {
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+impl CrdsTrie {
+    pub fn new() -> Self {
+        Self::Node(None, 0, None)
+    }
+
+    #[must_use]
+    pub fn insert(&mut self, value: &VersionedCrdsValue) -> bool {
+        self.insert_node(
+            CrdsFilter::hash_as_u64(&value.value_hash),
+            value.value.label(),
+        )
+    }
+
+    #[must_use]
+    pub fn remove(&mut self, value: &VersionedCrdsValue) -> bool {
+        self.remove_node(
+            CrdsFilter::hash_as_u64(&value.value_hash),
+            &value.value.label(),
+        )
+    }
+
+    /// Returns all the values where the mask_bits most significant bits of
+    /// their hash is the same as the given bit-mask.
+    pub fn find(&self, mask: u64, mask_bits: u32) -> impl Iterator<Item = &CrdsValueLabel> {
+        let node = self.find_node(mask, mask_bits);
+        CrdsTrieIterator::new(node).map(|v| v.iter()).flatten()
+    }
+
+    /// Traverses the trie following the bit pattern of the given hash value
+    /// and inserts the value at the final leaf node. Returns false if the trie
+    /// does not satisfy required invariants (which should indicate a bug).
+    fn insert_node(&mut self, hash: u64, value: CrdsValueLabel) -> bool {
+        let mut node = self;
+        let mut bit = MSB_U64;
+        while bit != 0 {
+            let branch = match node {
+                CrdsTrie::Leaf(_) => return false,
+                CrdsTrie::Node(left, size, right) => {
+                    *size += 1;
+                    if bit & hash == 0 {
```

Comment:
I think it would be good to extract this into a `left_or_right` function just because it's used in a lot of the other functions